base/inference.jl

```diff
@@ -2927,36 +2927,43 @@ function optimize(me::InferenceState)
 
     if isa(me.bestguess, Const) || isconstType(me.bestguess)
         me.const_ret = true
-        ispure = me.src.pure
-        if !ispure && length(me.src.code) < 10
-            ispure = true
+        proven_pure = false
+        # must be proven pure to use const_api; otherwise we might skip throwing errors
+        # (issue #20704)
+        if length(me.src.code) < 10
+            proven_pure = true
             for stmt in me.src.code
                 if !statement_effect_free(stmt, me.src, me.mod)
-                    ispure = false
+                    proven_pure = false
                     break
                 end
             end
-            if ispure
+            if proven_pure
                 for fl in me.src.slotflags
                     if (fl & Slot_UsedUndef) != 0
-                        ispure = false
+                        proven_pure = false
                         break
                     end
                 end
             end
         end
-        me.src.pure = ispure
+        if proven_pure
+            me.src.pure = true
+        end
 
-        do_coverage = coverage_enabled()
-        if ispure && !do_coverage
-            # use constant calling convention
-            # Do not emit `jlcall_api == 2` if coverage is enabled
-            # so that we don't need to add coverage support
-            # to the `jl_call_method_internal` fast path
-            # Still set pure flag to make sure `inference` tests pass
-            # and to possibly enable more optimization in the future
-            me.const_api = true
-            force_noinline || (me.src.inlineable = true)
+        if !coverage_enabled()
+            if proven_pure
+                # use constant calling convention
+                # Do not emit `jlcall_api == 2` if coverage is enabled
+                # so that we don't need to add coverage support
+                # to the `jl_call_method_internal` fast path
+                # Still set pure flag to make sure `inference` tests pass
+                # and to possibly enable more optimization in the future
+                me.const_api = true
+            end
+            if proven_pure || me.src.pure
```

Comment:
Ooops! and yes (merge with previous if statement)

It wasn't causing issues in the past since we also set `me.const_api`, so it should have managed to avoid actually inlining the whole function.